### PR TITLE
NH-93561: Target hibernate for trace context in queries

### PR DESCRIPTION
**Pull request by cleverchuk**
**Tl;dr**: Trace context in queries via hibernate class

**Context**:

This PR adds a new instrumentation to target `hibernate 6.x.x`. `DeferredResultSetAccess` class is the target of the instrumentation and the `excuteQuery` is the target method. This is done so that the generated span will be the parent of the execution span. This is achieved using reflection to access and wrap `statementCreator` field which is a Java8 `Function` interface that returns a `PreparedStatement` given a sql with an equivalent `FunctionWrapper`. The wrapper takes care of injecting the trace context.

Additionally, changes were made in how the unit tests are executed and some refactor to enable code sharing between modules. The unit tests will now run with the agent attached to validate spans created by instrumentation.

**Test Plan**:
Test services data [0](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1712644058766987264/overview?duration=21600), [1](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1712643928659124224/overview?duration=21600), [2](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1742334541200846848/logs?duration=21600) and [3](https://my.na-01.st-ssp.solarwinds.com/205939959869206528/entities/services/e-1777406072376840192/overview?duration=21600)

### CODE REVIEW

File: instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        @Advice.Local("swoScope") Scope scope) {
+      callDepth = CallDepth.forClass(DeferredResultSetAccess.class);
+      if (callDepth.getAndIncrement() > 0) {
+        return;
+      }
+
+      try {
+        // We're using reflection here because this seems to be the best place we can intercept the
+        // raw sql before the PreparedStatement is generated and be able to make the span a parent
+        // of the execution step.
```
Comment (cheempz) [inline instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java:75]: much appreciate the magic and the helpful comment above :) one question: the spans i've seen from testing are parametized and we don't collect actual argument values (which could contain sensitive info, good). but want to double check whether this would always be the case, e.g. is sanitization a concern for the `sql` value?

File: instrumentation/hibernate/hibernate-6.0/src/test/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentationTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  @Mock private JdbcCoordinator jdbcCoordinatorMock;
+
+  @Mock private LogicalConnectionImplementor logicalConnectionMock;
+
+  @Mock private ResourceRegistry resourceRegistryMock;
+
+  @Mock private SessionEventListenerManager sessionEventListenerManagerMock;
+
+  @Test
+  void verifyDbContextInjectionSpanIsCreated() {
```
Comment (cheempz) [inline instrumentation/hibernate/hibernate-6.0/src/test/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentationTest.java:70]: great to have this test! strangely not seeing it in the GHA workflow log output... should it be there? 

can the test also check for the `db.statement` attribute, and (not sure if meaningful given mocking) maybe sanitization of query literals.

File: instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+          statementCreator.set(drsa, new FunctionWrapper(function));
+        }
+        statementCreator.setAccessible(false);
+
+        context = HibernateInstrumenter.getInstance().start(parentContext, sql);
+        scope = context.makeCurrent();
+        swoSql = sql;
+
+      } catch (Throwable ignore) {
+        // ignore because we can't do jack here
```
Comment (cheempz) [inline instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java:105]: could log a message if it's safe and if the info could help diagnose instrumentation issue ;)

File: instrumentation/hibernate/hibernate-shared/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/HibernateInstrumenter.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+package com.solarwinds.opentelemetry.instrumentation.hibernate;
+
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+
+public final class HibernateInstrumenter {
+  private static final Instrumenter<String, Void> INSTANCE =
+      Instrumenter.<String, Void>builder(
+              GlobalOpenTelemetry.get(), "com.solarwinds.hibernate-6.0", (sql) -> "sw.jdbc.context")
```
Comment (cheempz) [inline instrumentation/hibernate/hibernate-shared/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/HibernateInstrumenter.java:25]: very nit, wonder if span name should be `sw.hibernate.context` instead.

File: instrumentation/jdbc/src/main/java/com/solarwinds/opentelemetry/instrumentation/StatementTruncator.java [hunk 5]
```diff
@@ -55,6 +55,14 @@ public static void maybeTruncateStatement(Context context) {
       int sqlMaxLength =
           ConfigManager.getConfigOptional(
               ConfigProperty.AGENT_SQL_QUERY_MAX_LENGTH, DEFAULT_SQL_MAX_LENGTH);
+
+      if (sql.contains("traceparent")) {
+        // Remove trace context from captured db.statement. For hibernate, we inject context before
+        // it's captured by upstream.
+        sql = sql.replaceAll("/.*traceparent.*/", "");
+        span.setAttribute(SemanticAttributes.DB_STATEMENT, sql);
+      }
```
Comment (cheempz) [inline instrumentation/jdbc/src/main/java/com/solarwinds/opentelemetry/instrumentation/StatementTruncator.java:64]: it's actually fine if we capture the injected comment as part of `db.statement` instead of a separate `QueryTag` attr.

**Review by cheempz — COMMENTED** [review]
Really appreciate all the iterations to arrive at this solution and this final well organized PR @cleverchuk! Left some comments :)

File: instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java (hunk 1 above)
Comment (cleverchuk) [reply instrumentation/hibernate/hibernate-6.0/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentation.java:75]: It'll be a concern. However, I have removed adding of the sql during changes for v4 since it's not needed for the for feature to work.

File: instrumentation/hibernate/hibernate-6.0/src/test/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentationTest.java (hunk 2 above)
Comment (cleverchuk) [reply instrumentation/hibernate/hibernate-6.0/src/test/java/com/solarwinds/opentelemetry/instrumentation/hibernate/v6_0/DrsaInstrumentationTest.java:70]: > great to have this test! strangely not seeing it in the GHA workflow log output... should it be there?

will be fixed in next push

> can the test also check for the db.statement attribute, and (not sure if meaningful given mocking) maybe sanitization of query literals.

won't necessary after the next push

File: instrumentation/hibernate/hibernate-shared/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/HibernateInstrumenter.java (hunk 4 above)
Comment (cleverchuk) [reply instrumentation/hibernate/hibernate-shared/src/main/java/com/solarwinds/opentelemetry/instrumentation/hibernate/HibernateInstrumenter.java:25]: sure

File: instrumentation/jdbc/src/main/java/com/solarwinds/opentelemetry/instrumentation/StatementTruncator.java (hunk 5 above)
Comment (cleverchuk) [reply instrumentation/jdbc/src/main/java/com/solarwinds/opentelemetry/instrumentation/StatementTruncator.java:64]: that increases the query length and could trigger truncation in the subsequent blocks which could after fingerprinting.

**Review by cheempz — APPROVED** [review]
LGTM! To recap, this PR also covers hibernate 4 and no longer captures `db.statement` in the generated span.
